### ISSUE TITLE
Question: How are min and max of cumulative variables assigned?

### ISSUE THREAD

**Issue by hklarner** [google/or-tools#761]
I created a routing problem and added some dimension to it. A solution assignment is found and I want to know the cumulative value at each index. I noticed that the `CumulVar` of an assignment does not only have a `Value` method but also `Min` and `Max` methods. Apparently the cumulative variables are implemented in such a way that they can represent intervals. I can see how setting

* `slack_max>0`
* `fix_start_cumul_to_zero=False`

introduces an ambiguity for the cumulative variables as their is a choice in how to start and how much slack to add at each stop. But

**Question**: How are the `Min` and `Max` at each index computed?

PS: Same question on stackoverflow: https://stackoverflow.com/questions/51371496/how-are-min-and-max-of-cumulative-variables-assigned, still figuring out where to post questions about or-tools

**Comment by Mizux** [comment]
First CumulVar is computed as follow:
 j == next(i), cumul(j) = cumul(i) + transit(i) + slack(i)
https://github.com/google/or-tools/blob/master/ortools/constraint_solver/routing.h#L383
without any constraints slack will be bound by zero to Max slack var (defined during the Dimension creation)

Now, for Time Windows, the solver use some propagators to determine the Min/Max at each node according to the previous and following node constraint(s) (i.e. TimeWindows constraints can shrink from  the slackVar=[0,MaxSlackVar] range)

ps: Usually questions should be post on the 
or-tools-discuss@googlegroups.com (this is barely the "only" source other team members regularly look at).
But hey! OSS outside world so SO and github issues is also (de facto) available so I try to look at them too..., note for github in an "unforeseeable" future we would like to keep it only for bug(s) and feature(s) request as an issue tracker should be used (i.e. move questions to SO/GoogleGroup :wink: or close it directly :disappointed:) .

**Comment by hklarner** [comment]
> j == next(i), cumul(j) = cumul(i) + transit(i) + slack(i)
Ok, this equation is helpful. So I tried confirming that it holds by printing for every index the cumulative value, the transit value and the slack value via

```
cumul = assignment.Value(time_dimension.CumulVar(i))
slack = assignment.Value(time_dimension.SlackVar(i))
```

If I set `fix_start_cumul_to_zero = True` then I can confirm the equation, sample output:

> 
> i: 0
> cumul(0): 0
> slack(0): 5
> transit(0,1): 10
> 
> i: 1
> cumul(1): 15
> slack(1): 0
> transit(1,2): 10
> 
> i: 2
> cumul(2): 25
> slack(2): 0
> transit(2,3): 10
>
> ...

 but if I set `fix_start_cumul_to_zero = False` then I numbers don't add up:

> i: 0
> cumul(0): 0
> slack(0): 0
> transit(0,1): 10
> 
> i: 1
> cumul(1): 15
> slack(1): 0
> transit(1,2): 10
> 
> i: 2
> cumul(2): 25
> slack(2): 0
> transit(2,3): 10
> 
> ...

Clearly a waiting time of `slack(0)=5` is implied here as well. To understand what is actually going on I print the Max and Min of each variable using

```
cumul_min = assignment.Min(time_dimension.CumulVar(i))
cumul_max = assignment.Max(time_dimension.CumulVar(i))
```

A sample output is then:

> i: 0
> cumul(0): [0,70]
> slack(0): [0,10]
> transit(0,1): 10
> 
> i: 1
> cumul(1): [15,80]
> slack(1): [0,10]
> transit(1,2): 10
> 
> i: 2
> cumul(2): [25,90]
> slack(2): [0,10]
> transit(2,3): 10
> 
> ...

But now it becomes trickier to interpret the assignment. cumul(0) lives in the interval [0,70] and slack(0) in the interval [0,10] but clearly there are dependencies between the choice of values, like cumul(0)=70 implies slack(0)=0 meaning I can not start late and have a large waiting time.

It seems to me that this affects also your example in https://developers.google.com/optimization/routing/cvrptw
Because what does "Total time of all routes: 376min" mean when the cumulative variable is in an interval? Shouldn't that imply that there is a fast and a slow solution? One that starts as late as possible with short waiting times and one that starts as early as possible with long waiting times?
